tidb-source-code-reading-21.md

```diff
@@ -0,0 +1,228 @@
+
+对于相关子查询，TiDB 会在 `expressionRewriter` 中将整个包含相关子查询的表达式转化为 `LogicalApply` 算子。`LogicalApply` 算子是一类特殊的 `LogicalJoin` ，特殊之处体现在执行逻辑上：对于 outer plan 返回的每一行记录，取出相关列的具体值传递给子查询，再执行根据子查询生成的 inner plan ，即 `LogicalApply` 在执行时只能选择类似 循环嵌套连接的方式，而普通的 `LogicalJoin` 则可以在物理优化阶段根据代价模型选择最合适的执行方式，包括 `HashJoin`，`MergeJoin` 和 `IndexLookUpJoin`，理论上后者生成的物理执行计划一定会比前者更优，所以在逻辑优化阶段我们会检查是否可以应用“去相关”这一优化规则，试图将 `LogicalApply` 转化为等价的 `LogicalJoin` 。其核心思想是将 `LogicalApply` 的 inner plan 中包含相关列的那些算子提升到 `LogicalApply` 之中或之上，在算子提升后如果 inner plan 中不再包含任何的相关列，即不再引用任何 outer plan 中的列，那么 `LogicalApply` 就会被转换为普通的 `LogicalJoin` ，这部分代码逻辑实现在 [rule_decorrelate.go](https://github.com/eurekaka/tidb/blob/logical_rules_reading/planner/core/rule_decorrelate.go) 文件中。
+
+具体的算子提升方式分为以下几种情况：
+
+* **inner plan 的根节点是 `LogicalSelection`**
+
+    则将其过滤条件添加到 `LogicalApply` 的 join condition 中，然后将该 `LogicalSelection` 从 inner plan 中删除，再递归地对 inner plan 提升算子。
+
+    以如下查询为例：
+
+    ```sql
+    select * from t1 where t1.a in (select t2.a from t2 where t2.b = t1.b);
+    ```
+
+    其生成的最初执行计划片段会是：
+
+    ![1.png](https://upload-images.jianshu.io/upload_images/542677-d01f93be2416556f.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
+
+    `LogicalSelection` 提升后会变成如下片段：
+
+    ![2.png](https://upload-images.jianshu.io/upload_images/542677-7d11730db1688b53.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
+
+    到此 inner plan 中不再包含相关列，于是 `LogicalApply` 会被转换为如下 LogicalJoin ：
+
+    ![3.png](https://upload-images.jianshu.io/upload_images/542677-9b7df1ed09ba348d.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
+
+* **inner plan 的根节点是 `LogicalMaxOneRow`**
+
+    即要求子查询最多输出一行记录，比如这个例子：
+
+    ```sql
+    select *, (select t2.a from t2 where t2.pk = t1.a) from t1;
+    ```
+
+    因为子查询出现在整个查询的投影项里，所以 `expressionRewriter` 在处理子查询时会对其生成的执行计划在根节点上加一个 `LogicalMaxOneRow` 限制最多产生一行记录，如果在执行时发现下层输出多于一行记录，则会报错。在这个例子中，子查询的过滤条件是 `t2` 表的主键上的等值条件，所以子查询肯定最多只会输出一行记录，而这个信息在“构建节点属性”这一步时会被发掘出来并记录在算子节点的 MaxOneRow 属性中，所以这里的 `LogicalMaxOneRow` 节点实际上是冗余的，于是我们可以将其从 inner plan 中移除，然后再递归地对 inner plan 做算子提升。
+
+* **inner plan 的根节点是 `LogicalProjection`**
+
+    则首先将这个投影算子从 inner plan 中移除，再根据 `LogicalApply` 的连接类型判断是否需要在 `LogicalApply` 之上再加上一个 `LogicalProjection` ，具体来说是：对于非 semi-join 这一类的连接（包括 inner join 和 left join ），inner plan 的输出列会保留在 `LogicalApply` 的结果中，所以这个投影操作需要保留，反之则不需要。最后，再递归地对删除投影后的 inner plan 提升下层算子。
+
+* **inner plan 的根节点是 `LogicalAggregation`**
+
+1. 首先我们会检查这个聚合算子是否可以被提升到 `LogicalApply` 之上再执行。以如下查询为例：
+
+    ```sql
+    select *, (select sum(t2.b) from t2 where t2.a = t1.pk) from t1;
+    ```
+
+    其最初生成的执行计划片段会是：
+
+    ![4.png](https://upload-images.jianshu.io/upload_images/542677-c812770e277e0a4a.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
```

Comment:
The picture is not complete?

Reply:
These pictures  sometimes didn't load completely  until the PR  was merged. Maybe there are some bugs with markdown preview...but it's fine, just ignore this problem.